iothub/device/src/Logger/EventCounterLogger.cs

```diff
@@ -0,0 +1,347 @@
+        /// </summary>
+        void OnAmqpTokenRefreshed();
+        /// <summary>
+        /// Enable logger and start logging
+        /// </summary>
+        /// <param name="interval">The interval of each logging period.</param>
+        /// <param name="redirectToConsole">Dump log to file.</param>
+        /// <param name="cancellationToken">Cancel the loop.</param>
+        Task StartLoggerAsync(TimeSpan interval, bool redirectToConsole, CancellationToken cancellationToken);
+    }
+
+#if NETSTANDARD2_0
+    /// <summary>
+    /// Logger for device client events.
+    /// </summary>
+    [EventSource(Name = "Microsoft-Azure-Devices-Client-Logger-Event-Counter")]
+    public class EventCounterLogger : EventSource, IEventCounterLogger
+    {
+        private static EventCounterLogger s_instance = new EventCounterLogger();
+
+        
+
+        private bool _started;
+        private int _deviceClientCreationCounts;
+        private int _deviceClientDisposeCounts;
+        private int _amqpUnitCreationCounts;
+        private int _amqpUnitDisposeCounts;
+        private int _amqpConnectionEstablishCounts;
+        private int _amqpConnectionDisconnectionCounts;
+        private int _amqpSessionEstablishCounts;
+        private int _amqpSessionDisconnectionCounts;
+        private int _amqpTokenRefreshStartCounts;
+        private int _amqpTokenRefreshStopCounts;
+        private int _amqpTokenRefreshCounts;
+
+        /// <summary>
+        /// Return the instance.
+        /// </summary>
+        public static IEventCounterLogger GetInstance()
+        {
+            return s_instance;
+        }
+
+        private EventCounterLogger() : base()
+        {
+        }
+
+        /// <summary>
+        /// When device client was created.
+        /// </summary>
+        public void OnDeviceClientCreated()
+        {
+            if (_started)
+            {
+                Interlocked.Increment(ref _deviceClientCreationCounts);
+            }
+        }
+
+        /// <summary>
+        /// When device client was disposed.
+        /// </summary>
+        public void OnDeviceClientDisposed()
+        {
+            if (_started)
+            {
+                Interlocked.Increment(ref _deviceClientDisposeCounts);
+            }
+        }
+
+        /// <summary>
+        /// When Amqp unit was created.
+        /// </summary>
+        public void OnAmqpUnitCreated()
+        {
+            if (_started)
+            {
+                Interlocked.Increment(ref _amqpUnitCreationCounts);
+            }
+        }
+
+        /// <summary>
+        /// When Amqp unit was disposed.
+        /// </summary>
+        public void OnAmqpUnitDisposed()
+        {
+            if (_started)
+            {
+                Interlocked.Increment(ref _amqpUnitDisposeCounts);
+            }
+        }
+
+        /// <summary>
+        /// When AMQP connection was established.
+        /// </summary>
+        public void OnAmqpConnectionEstablished()
```

Comment:
These are AMQP specific - what is the strategy to add the same counters for all other transports? Should we have common ones under the same counter (e.g. # of active connections?)